shell/browser/electron_api_ipc_handler_impl.cc

```diff
@@ -38,57 +43,129 @@ void ElectronApiIPCHandlerImpl::OnConnectionError() {
                                        InvokeCallback callback) {
-  api::WebContents* api_web_contents = api::WebContents::From(web_contents());
-  if (api_web_contents) {
-    api_web_contents->Invoke(internal, channel, std::move(arguments),
-                             std::move(callback), GetRenderFrameHost());
-  }
+  auto* session = GetSession();
+  v8::Isolate* isolate = electron::JavascriptEnvironment::GetIsolate();
+  v8::HandleScope handle_scope(isolate);
+  auto event = MakeIPCEvent(isolate, session, internal, std::move(callback));
+  if (event.IsEmpty())
+    return;
+  session->Invoke(event, channel, std::move(arguments));
 }
 
 void ElectronApiIPCHandlerImpl::ReceivePostMessage(
     const std::string& channel,
     blink::TransferableMessage message) {
-  api::WebContents* api_web_contents = api::WebContents::From(web_contents());
-  if (api_web_contents) {
-    api_web_contents->ReceivePostMessage(channel, std::move(message),
-                                         GetRenderFrameHost());
-  }
+  auto* session = GetSession();
+  v8::Isolate* isolate = electron::JavascriptEnvironment::GetIsolate();
+  v8::HandleScope handle_scope(isolate);
+  auto event = MakeIPCEvent(isolate, session, false);
+  if (event.IsEmpty())
+    return;
+  session->ReceivePostMessage(event, channel, std::move(message));
 }
 
 void ElectronApiIPCHandlerImpl::MessageSync(bool internal,
                                             const std::string& channel,
                                             blink::CloneableMessage arguments,
                                             MessageSyncCallback callback) {
-  api::WebContents* api_web_contents = api::WebContents::From(web_contents());
-  if (api_web_contents) {
-    api_web_contents->MessageSync(internal, channel, std::move(arguments),
-                                  std::move(callback), GetRenderFrameHost());
-  }
+  auto* session = GetSession();
+  v8::Isolate* isolate = electron::JavascriptEnvironment::GetIsolate();
+  v8::HandleScope handle_scope(isolate);
+  auto event = MakeIPCEvent(isolate, session, internal, std::move(callback));
+  if (event.IsEmpty())
+    return;
+  session->MessageSync(event, channel, std::move(arguments));
 }
 
 void ElectronApiIPCHandlerImpl::MessageHost(const std::string& channel,
                                             blink::CloneableMessage arguments) {
+  auto* session = GetSession();
+  v8::Isolate* isolate = electron::JavascriptEnvironment::GetIsolate();
+  v8::HandleScope handle_scope(isolate);
+  auto event = MakeIPCEvent(isolate, session, false);
+  if (event.IsEmpty())
+    return;
   api::WebContents* api_web_contents = api::WebContents::From(web_contents());
-  if (api_web_contents) {
-    api_web_contents->MessageHost(channel, std::move(arguments),
-                                  GetRenderFrameHost());
-  }
+  api_web_contents->MessageHost(event, channel, std::move(arguments));
```

Comment:
what caused this to be not routed via session ?

Reply:
MessageHost is only used for webviews which has no equivalent for service workers. I figured we could be more direct with emitting that event.

Reply:
I can certainly change it if desired. 

Currently it emits on WebContents.
After it will emit on Session then re-emit on WebContents.

Adds some minor overhead, but does keep the dispatching and code more consistent.

Reply:
The current change makes other channels emit from session to re-emit on webContents, the concern for this particular channel is that it has to re-emit again on the embedder webContents ? Can we measure the overhead which can help decide the preferred path.